operator/es_client.go

```diff
@@ -67,22 +68,68 @@ type ESHealth struct {
 	Status string `json:"status"`
 }
 
+type Exclude struct {
+	IP null.String `json:"_ip,omitempty"`
+}
+
+type Allocation struct {
+	Exclude Exclude `json:"exclude,omitempty"`
+}
+type Rebalance struct {
+	Enable null.String `json:"enable,omitempty"`
+}
+
+type Routing struct {
+	Allocation Allocation `json:"allocation,omitempty"`
+	Rebalance  Rebalance  `json:"rebalance,omitempty"`
+}
+
+type Cluster struct {
+	Routing Routing `json:"routing,omitempty"`
+}
+
+type ClusterSettings struct {
+	Cluster Cluster `json:"cluster"`
+}
+
 // ESSettings represent response from _cluster/settings
 type ESSettings struct {
-	Transient struct {
-		Cluster struct {
-			Routing struct {
-				Allocation struct {
-					Exclude struct {
-						IP string `json:"_ip"`
-					} `json:"exclude"`
-				} `json:"allocation"`
-				Rebalance struct {
-					Enable string `json:"enable"`
-				} `json:"rebalance"`
-			} `json:"routing"`
-		} `json:"cluster"`
-	} `json:"transient"`
+	Transient  ClusterSettings `json:"transient,omitempty"`
+	Persistent ClusterSettings `json:"persistent,omitempty"`
+}
+
+func (esSettings *ESSettings) MergeNonEmptyTransientSettings() {
+	if value := esSettings.GetTransientRebalance().ValueOrZero(); value != "" {
+		esSettings.Persistent.Cluster.Routing.Rebalance.Enable = null.StringFromPtr(&value)
+		esSettings.Transient.Cluster.Routing.Rebalance.Enable = null.StringFromPtr(nil)
+	}
+
+	if transientExcludeIps := esSettings.GetTransientExcludeIPs().ValueOrZero(); transientExcludeIps != "" {
+		persistentExcludeIps := esSettings.GetPersistentExcludeIPs().ValueOrZero()
+		if persistentExcludeIps == "" {
+			esSettings.Persistent.Cluster.Routing.Allocation.Exclude.IP = null.StringFromPtr(&transientExcludeIps)
+		} else {
+			mergedIps := null.StringFrom(transientExcludeIps + "," + persistentExcludeIps)
```

Comment:
do we need to deduplicate?

Reply:
I don't think it's strictly required for the es-operator manipulations because the new exclude ip list will also have this duplicated entry. It might cause some confusion if for some reason a developer interfered and is manually updating the ip exclusion list and doesn't notice the duplicate entry. It's better to deduplicate to avoid such edge cases.